core/emulate.c

```diff
@@ -0,0 +1,908 @@
+    X8(F(em_dec, op_modrm_reg, op_none, op_none, 0)),
+    /* 0x50 - 0x7F */
+    X16(N), X16(N), X16(N),
+    /* 0x80 - 0x8F */
+    G(opcode_group1, op_modrm_rm, op_imm, op_none, INSN_BYTEOP),
+    G(opcode_group1, op_modrm_rm, op_imm, op_none, 0),
+    G(opcode_group1, op_modrm_rm, op_imm, op_none, INSN_BYTEOP),
+    G(opcode_group1, op_modrm_rm, op_simm8, op_none, 0),
+    X4(N),
+    I2_BV(em_mov, op_modrm_rm, op_modrm_reg, op_none, INSN_MODRM | INSN_MOV),
+    I2_BV(em_mov, op_modrm_reg, op_modrm_rm, op_none, INSN_MODRM | INSN_MOV),
+    X4(N),
+    /* 0x90 - 0x9F */
+    X16(N),
+    /* 0xA0 - 0xAF */
+    I2_BV(em_mov, op_acc, op_moffs, op_none, INSN_MOV),
+    I2_BV(em_mov, op_moffs, op_acc, op_none, INSN_MOV),
+    I2_BV(em_mov, op_di, op_si, op_none, INSN_MOV | INSN_REP), /* movs{b,w,d,q} */
+    X2(N),
+    X2(N),
+    I2_BV(em_mov, op_di, op_acc, op_none, INSN_MOV | INSN_REP), /* stos{b,w,d,q} */
+    I2_BV(em_mov, op_acc, op_si, op_none, INSN_MOV | INSN_REP), /* lods{b,w,d,q} */
+    X2(N),
+    /* 0xB0 - 0xBF */
+    X16(N),
+    /* 0xC0 - 0xCF */
+    X4(N),
+    X2(N),
+    G(opcode_group11, op_modrm_rm, op_imm, op_none, INSN_BYTEOP),
+    G(opcode_group11, op_modrm_rm, op_imm, op_none, 0),
+    X8(N),
+    /* 0xD0 - 0xEF */
+    X16(N), X16(N),
+    /* 0xF0 - 0xFF */
+    X4(N),
+    X2(N),
+    G(opcode_group3, op_none, op_none, op_none, INSN_BYTEOP),
+    G(opcode_group3, op_none, op_none, op_none, 0),
+    X8(N),
+};
+
+static const struct em_opcode_t opcode_table_0F[256] = {
+    /* 0x00 - 0xAF */
+    X16(N), X16(N), X16(N), X16(N),
+    X16(N), X16(N), X16(N), X16(N),
+    X16(N), X16(N), X16(N),
+    /* 0xB0 - 0xBF */
+    X4(N),
+    X2(N),
+    I2_BV(em_movzx, op_modrm_reg, op_modrm_rm, op_none, INSN_MODRM | INSN_MOV),
+    X8(N),
+    /* 0xC0 - 0xFF */
+    X16(N), X16(N), X16(N), X16(N),
+};
+
+static const struct em_opcode_t opcode_table_0F38[256] = {
+    /* 0x00 - 0xFF */
+    X16(N), X16(N), X16(N), X16(N),
+    X16(N), X16(N), X16(N), X16(N),
+    X16(N), X16(N), X16(N), X16(N),
+    X16(N), X16(N), X16(N), X16(N),
+};
+
+static const struct em_opcode_t opcode_table_0F3A[256] = {
+    /* 0x00 - 0xFF */
+    X16(N), X16(N), X16(N), X16(N),
+    X16(N), X16(N), X16(N), X16(N),
+    X16(N), X16(N), X16(N), X16(N),
+    X16(N), X16(N), X16(N), X16(N),
+};
+
+/* Emulate accesses to guest memory */
+static uint64_t get_canonical_address(struct em_context_t *ctxt,
+                                      uint64_t addr, uint vaddr_bits)
+{
+    return ((int64)addr << (64 - vaddr_bits)) >> (64 - vaddr_bits);
+}
+
+static em_status_t get_linear_address(struct em_context_t *ctxt,
+                                      struct operand_mem_t *mem,
+                                      uint64_t *la)
+{
+    uint64_t linear;
+
+    linear = ctxt->ops->get_segment_base(ctxt->vcpu, mem->seg) + mem->ea;
+    if (ctxt->mode == EM_MODE_PROT64) {
+        linear = get_canonical_address(ctxt, linear, 48);
+    } else {
```

Comment:
Add a case for `EM_MODE_REAL`, where `linear` should be truncated to 20 bits. See Intel SDM Vol. 1, 3.3.5 _32-Bit and 16-Bit Address and Operand Sizes_:

> An address-size override can be used in real-address mode to enable 32-bit addressing. However, the maximum allowable 32-bit linear address is still 000FFFFFH (2^20 - 1).